app/controllers/shared/api.php

```diff
@@ -208,19 +208,52 @@
         $useCache = $route->getLabel('cache', false);
 
         if ($useCache) {
-            $key = md5($request->getURI() . implode('*', $request->getParams()));
+            $key = md5($request->getURI() . implode('*', $request->getParams())) . '*' . APP_CACHE_BUSTER;
             $cache = new Cache(
                 new Filesystem(APP_STORAGE_CACHE . DIRECTORY_SEPARATOR . 'app-' . $project->getId())
             );
             $timestamp = 60 * 60 * 24 * 30;
             $data = $cache->load($key, $timestamp);
+
             if (!empty($data)) {
                 $data = json_decode($data, true);
+                $parts = explode('/', $data['resourceType']);
+                $type = $parts[0] ?? null;
+
+                if ($type === 'bucket') {
+                    $bucketId = $parts[1] ?? null;
+
+                    $bucket = Authorization::skip(fn () => $dbForProject->getDocument('buckets', $bucketId));
+
+                    if ($bucket->isEmpty() || (!$bucket->getAttribute('enabled') && $mode !== APP_MODE_ADMIN)) {
+                        throw new Exception(Exception::STORAGE_BUCKET_NOT_FOUND);
+                    }
+
+                    $fileSecurity = $bucket->getAttribute('fileSecurity', false);
+                    $validator = new Authorization(Database::PERMISSION_READ);
+                    $valid = $validator->isValid($bucket->getRead());
+                    if (!$fileSecurity && !$valid) {
+                        throw new Exception(Exception::USER_UNAUTHORIZED);
+                    }
+
+                    $parts = explode('/', $data['resource']);
+                    $fileId = $parts[1] ?? null;
+
+                    if ($fileSecurity && !$valid) {
+                        $file = $dbForProject->getDocument('bucket_' . $bucket->getInternalId(), $fileId);
+                    } else {
+                        $file = Authorization::skip(fn() => $dbForProject->getDocument('bucket_' . $bucket->getInternalId(), $fileId));
+                    }
+
+                    if ($file->isEmpty()) {
+                        throw new Exception(Exception::STORAGE_FILE_NOT_FOUND);
+                    }
```

Comment:
I tested and confirm this does work. It's just unfortunate there's a lot of duplicate code between here and the storage API.